django/db/models/functions/comparison.py

```diff
@@ -158,13 +160,21 @@ def as_sql(self, compiler, connection, **extra_context):
         return super().as_sql(compiler, connection, **extra_context)
 
     def as_postgresql(self, compiler, connection, **extra_context):
-        return self.as_sql(
+        sql, params = self.as_sql(
             compiler,
             connection,
             function="JSONB_BUILD_OBJECT",
             **extra_context,
         )
 
+        # Enforce str to be text rather than unknown, otherwise they cannot
+        # be merged server-side
+        for i, p in enumerate(params):
+            if isinstance(p, str):
+                params[i] = Text(p)
```

Comment:
this pattern can be written slightly faster as a list comp:

```suggestion
    params = [Text(p) if instance(p, str) else p for p in params]
```

i checked with ipython:

```
In [6]: params = [str(x) for x in range(1000)]

In [7]: %%timeit
   ...: for i, p in enumerate(params):
   ...:     if isinstance(p, str):
   ...:         params[i] = p
   ...:
165 µs ± 1.11 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [8]: %timeit [p if isinstance(p, str) else p for p in params]
74.3 µs ± 773 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)
```

(and for small lists too)

Reply:
Good point, I have updated it. I didn't look for performance gains at all yet since I was mainly concerned with getting the errors fixed :)